text/0000-macro-expansion-for-macro-input.md

```diff
@@ -0,0 +1,214 @@
+
+Thankfully, there's a way to _tell_ Rust to treat some tokens as macros, and to expand them before trying to expand _your_ macro. But first, we need to understand how Rust finds and expands macros.
+
+## Macro expansion and marking
+
+Rust uses an iterative process to expand macros. The loop that Rust performs is roughly as follows:
+
+1. Look for and expand any macros we can parse in expression, item, or attribute position.
+    - Skip any macros that have explicitly marked macros in their arguments.
+    - Are there any new macros we can parse and expand? Go back to step 1.
+2. Look for and expand any explicitly marked macros where there are raw tokens (like inside the arguments to a proc macro or attribute macro).
+    - Are there any new explicitly marked macros we can expand? Go back to step 2.
+    - Otherwise, go back to step 1.
+
+Other than some details about handing macros we can't resolve yet (maybe because they're defined by another macro expansion), that's it!
+
+In order to explicitly mark a macro for the compiler to expand, we actually just mark the `!` or `#` token on the macro call. The compiler looks around the token for the other bits it needs.
+
+In most cases, when you're writing a proc or attribute macro you don't really need that level of precision when marking macros. Instead, you just want to expand every macro in your input before continuing!
+
+The `syn` crate has a utility attribute `#[syn(expand_input)]` which converts a normal proc or attribute macro into one that does that expansion. For example, if we add `#[syn(expand_input)]` to our `string_length` proc macro above, we get something like:
+
+
+```rust
+#[proc_macro]
+pub fn string_length(tokens: TokenStream) -> TokenStream {
+    if let Some(marked_tokens) = syn::find_and_mark_all_macros(&tokens) {
+        return quote!(string_length!(#marked_tokens));
+    }
+    // Otherwise, continue as before.
+    ...
+}
+
+```
+
+Notice that in the `quote!` output, the _argument_ to the new call to `string_length!` is marked by `syn::find_and_mark_all_macros`, but the _new call itself_ is unmarked. Recalling the macro expansion process we outlined earlier, that means the arguments will all get expanded before `string_length!` gets expanded again (hopefully without any macros in the arguments, but if there are then this whole process just repeats).
+
+# Reference-level explanation
+
+Currently, the compiler does actually perform something similar to the loop described in the section on [expansion order](#macro-expansion-and-marking). We could 'just' augment the step that identifies potential macro calls to also inspect the otherwise unstructured token trees within macro arguments.
+
+This proposal requires that some tokens contain extra semantic information similar to the existing `Span` API. Since that API (and its existence) is in a state of flux, details on what this 'I am a macro call that you need to expand!' idea may need to wait until those have settled.
```

Comment:
The implementation may also choose to put the marks elsewhere, as long as the mark gets followed up.

Reply:
I'm not sure what point you're making - is it that the new token implementation might not have token attributes attached to tokens at all, and instead have some sort of registry as the public API?